File: kirun-java/src/main/java/com/fincity/nocode/kirun/engine/model/Position.java
package com.fincity.nocode.kirun.engine.model;

import java.io.Serializable;
import java.util.Map;

import com.fincity.nocode.kirun.engine.json.schema.Schema;
import com.fincity.nocode.kirun.engine.json.schema.type.SchemaType;
import com.fincity.nocode.kirun.engine.json.schema.type.Type;
import com.fincity.nocode.kirun.engine.namespaces.Namespaces;

import lombok.Data;

@Data
public class Position implements Serializable {

	private static final long serialVersionUID = -7667086732611242476L;

	public static final String SCHEMA_NAME = "Position";

	public static final Schema SCHEMA = new Schema().setNamespace(Namespaces.SYSTEM)
	        .setName(SCHEMA_NAME)
	        .setType(Type.of(SchemaType.OBJECT))
	        .setProperties(Map.of("left", Schema.ofFloat("left"), "top", Schema.ofFloat("top")));

	private float left;
	private float top;
}

File: kirun-js/src/engine/model/Position.ts
import { Schema } from '../json/schema/Schema';
import { SchemaType } from '../json/schema/type/SchemaType';
import { TypeUtil } from '../json/schema/type/TypeUtil';
import { Namespaces } from '../namespaces/Namespaces';

export class Position {
    public static readonly SCHEMA_NAME: string = 'Position';
    public static readonly SCHEMA: Schema = new Schema()
        .setNamespace(Namespaces.SYSTEM)
        .setName(Position.SCHEMA_NAME)
        .setType(TypeUtil.of(SchemaType.OBJECT))
        .setProperties(
            new Map([
                ['left', Schema.ofFloat('left')],
                ['top', Schema.ofFloat('top')],
            ]),
        );
    private left: number;
    private top: number;

    public constructor(left: number, top: number) {
        this.left = left;
        this.top = top;
    }

    public getLeft(): number {
        return this.left;
    }
    public setLeft(left: number): Position {
        this.left = left;
        return this;
    }
    public getTop(): number {
        return this.top;
    }
    public setTop(top: number): Position {
        this.top = top;
        return this;
    }

    public static from(json: any): Position {
        if (!json) return new Position(-1, -1);
        return new Position(json.left, json.top);
    }
}
